### DISCUSSION TITLE
How to check for collision with widget offsets?

### DISCUSSION THREAD

**Discussion by TomJGooding** [Textualize/textual#4065] [Q&A]
I'm sure I am missing something obvious, but how would you check for collisions when setting widget offsets?

Here's a simple example where the falling bars should stack on top of each other when they fall. I've tried `get_widget_at` and other methods but I can't seem to get this to work. Thanks in advance for any advice!

```python
from itertools import cycle

from textual.app import App
from textual.widgets import Static

COLORS = cycle(["red", "orange", "yellow", "green", "blue", "purple"])


class Bar(Static):
    def on_mount(self) -> None:
        self.styles.background = next(COLORS)
        self.timer = self.set_interval(0.05, self.fall)

    def fall(self) -> None:
        x = int(self.styles.offset.x.value)
        y = int(self.styles.offset.y.value)

        screen_height = self.screen.size.height
        collision = False  # TODO: check if aleady a bar at y + 1

        if y + 1 >= screen_height or collision:
            self.timer.stop()
            return
        self.styles.offset = (x, y + 1)


class FallingExample(App):
    BINDINGS = [("space", "add_bar")]

    def action_add_bar(self) -> None:
        self.mount(Bar())


if __name__ == "__main__":
    app = FallingExample()
    app.run()
```

**Comment by willmcgugan** [comment]
`get_widget_at` should work. But bear in mind it takes screen coordinates.

I'm assuming this is for a game? I'd probably implement your own collision detection. The code in geometry.py will get you part of the way there...

**Reply by TomJGooding** [reply]
Thanks WIll. Perhaps I've misunderstood the concept of `offset` as nothing I've tried seems to work. If you have any spare time* to correct my example code to demonstrate how to use `get_widget_at`, it would be much appreciated!

<sub>* Obviously I appreciate you don't have much spare time as an open-source maintainer!</sub>

**Reply by davep** [reply]
I'm not Will, but... Won't there be a problem here if you're using `get_widget_at` along with the offset to calculate the collision? I'm guessing you'd be doing something akin to this:

```python
from itertools import cycle

from textual.app import App
from textual.errors import NoWidget
from textual.widgets import Static

COLORS = cycle(["red", "orange", "yellow", "green", "blue", "purple"])


class Bar(Static):

    DEFAULT_CSS = """
    Bar { height: 1; }
    """

    def on_mount(self) -> None:
        self.styles.background = next(COLORS)
        self.call_after_refresh(self.start_timer)

    def start_timer(self) -> None:
        self.timer = self.set_interval(0.05, self.fall)

    def fall(self) -> None:
        x = int(self.styles.offset.x.value)
        y = int(self.styles.offset.y.value)

        screen_height = self.screen.size.height
        try:
            widget, _ = self.screen.get_widget_at(0, y+1)
            self.update(f"Found {widget} which is {widget.styles.background} - is me: {widget == self}")
            collision = isinstance(widget, Bar)
        except NoWidget:
            collision = False
            self.update("NoWidget")

        if y + 1 >= screen_height or collision:
            self.timer.stop()
            return
        self.styles.offset = (x, y + 1)


class FallingExample(App):
    BINDINGS = [("space", "add_bar")]

    def action_add_bar(self) -> None:
        self.mount(Bar())


if __name__ == "__main__":
    app = FallingExample()
    app.run()
```

When I try this out the second bar I add collides with itself. This seems to make sense because the second bar will be mounted at Y=1 (the first is still mounted at Y=0, it just happens to be visually offset, but it's still mounted at 0 and so pushes the second one down to 1), then its code will work out where the next row is by looking at its Y offset plus one; when it starts out its Y offset will be 0, so it'll look for a `Bar` at 0+1. which is Y=1. So it'll find itself and conclude that it's collided with itself.

**Reply by davep** [reply]
PS: Ignore the faff with `call_after_refresh` and so on; I had a moment of wondering if it was a subtle timing issue before recognising what might really be going on, and the above is just the state of my sandbox code including that dead-end-only-one-coffee-into-the-day thought.

**Reply by TomJGooding** [reply]
> This seems to make sense because the second bar will be mounted at Y=1 (the first is still mounted at Y=0, it just happens to be visually offset, but it's still mounted at 0 and so pushes the second one down to 1), then its code will work out where the next row is by looking at its Y offset plus one; when it starts out its Y offset will be 0, so it'll look for a `Bar` at 0+1. which is Y=1.

You wrote this out after only coffee?!

I _think_ I understand...? After experimenting with `self.mount(Bar(), before=0)` I certainly how the bars are pushed down when new ones are added.
